plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/ApplyWorkspaceEditAction.java

```diff
@@ -147,6 +188,141 @@ public void applyWorkspaceEdit(WorkspaceEdit edit) {
+      return promises.resolve(null);
+    }
+
+    ArrayOf<Promise<?>> changesPromises = elemental.util.Collections.arrayOf();
+
+    for (Either<ResourceChange, TextDocumentEdit> rChange : resourceChanges) {
+      if (rChange.isRight()) {
+        continue;
+      }
+
+      ResourceChange change = rChange.getLeft();
+      if (change == null) {
+        continue;
+      }
+      String newUri = change.getNewUri();
+      String current = change.getCurrent();
+
+      Path path = toPath(newUri).removeFirstSegments(1).makeAbsolute();
+      if (isNullOrEmpty(current)) {
+        createResource(path, notification);
+        continue;
+      }
+      Path oldPath = toPath(current).removeFirstSegments(1).makeAbsolute();
+
+      Container workspaceRoot = appContext.getWorkspaceRoot();
+      changesPromises.push(
+          workspaceRoot
+              .getResource(oldPath)
+              .then(
+                  arg -> {
+                    if (!arg.isPresent()) {
+                      return;
+                    }
+
+                    Resource resource = arg.get();
+
+                    editorAgent.saveAll(
+                        new AsyncCallback<Void>() {
+                          @Override
+                          public void onFailure(Throwable throwable) {
+                            notification.setContent("Can't save files.");
+                          }
+
+                          @Override
+                          public void onSuccess(Void aVoid) {
+                            final List<EditorPartPresenter> openedEditors =
+                                editorAgent.getOpenedEditors();
+                            for (EditorPartPresenter editor : openedEditors) {
+                              if (resource
+                                  .getLocation()
+                                  .isPrefixOf(editor.getEditorInput().getFile().getLocation())) {
+                                TextDocumentIdentifier documentId =
+                                    dtoHelper.createTDI(editor.getEditorInput().getFile());
+                                DidCloseTextDocumentParams closeEvent =
+                                    dtoFactory.createDto(DidCloseTextDocumentParams.class);
+                                closeEvent.setTextDocument(documentId);
+                                textDocumentService.didClose(closeEvent);
+                              }
+                            }
+                            moveResource(resource, path, notification);
+                          }
+                        });
+                  }));
+    }
+    return promises.all2(changesPromises).thenPromise(ignored -> promises.resolve(null));
+  }
+
+  private void createResource(Path path, Notification notification) {
+    Container workspaceRoot = appContext.getWorkspaceRoot();
+    workspaceRoot
+        .getResource(path)
+        .then(
+            resource -> {
+              if (isNullOrEmpty(path.getFileExtension())) {
```

Comment:
I dont' think this is a valid condition. It's perfectly valid to create files without extensions and folders with.

Reply:
yes, but I don't see another way  to recognize which type of resource is here. Is it a folder or is a file.